File: F#/chap07/Chap0704.fs
namespace Chap0704

module C070401 =
    type ProductCode = Undefined
    type ValidatedOrder = Undefined

    type CheckProductCodeExits = ProductCode -> bool

    type UnvalidatedAddress = Undefined
    type CheckedAddress = CheckedAddress of UnvalidatedAddress
    type AddressValidationError = AddressValidationError of string
    type CheckAddressExists = UnvalidatedAddress -> Result<CheckedAddress, AddressValidationError>

    type ValidationError = Undefined
    type UnvalidatedOrder = Undefined

    type ValidateOrder =
        CheckProductCodeExits // 依存関係
            -> CheckAddressExists // 依存関係
            -> UnvalidatedOrder // 入力
            -> Result<ValidatedOrder, ValidationError> // 出力

module C070402 =
    open C070401

    type Price = Undefined
    type PricedOrder = Undefined
    type GetProductPrice = ProductCode -> Price

    type PriceOrder =
        GetProductPrice // 依存関係
            -> ValidatedOrder // 入力
            -> PricedOrder // 出力


module C070403 =
    open C070402

    type EmailAddress = Undefined
    type HtmlString = HtmlString of string

    type OrderAcknowledgment =
        { EmailAddress: EmailAddress
          Letter: HtmlString }

    type CreateOrderAcknowledgmentLetter = PricedOrder -> HtmlString

    type SendOrderAcknowledgmentRetVoid = OrderAcknowledgment -> unit
    type SendOrderAcknowledgmentRetBool = OrderAcknowledgment -> bool

    type SendResult =
        | Sent
        | NotSent

    type SendOrderAcknowledgment = OrderAcknowledgment -> SendResult

    type OrderId = Undefined

    type OrderAcknowledgmentSent =
        { OrderId: OrderId
          EmailAddress: EmailAddress }

    type SendOrderAcknowledgmentRetEvent = OrderAcknowledgment -> OrderAcknowledgmentSent option

    type AcknowledgementOrder =
        CreateOrderAcknowledgmentLetter // 依存関係
            -> SendOrderAcknowledgment // 依存関係
            -> PricedOrder // 入力
            -> OrderAcknowledgmentSent option // 出力


module C070404 =
    open chap0702
    open C070403

    type Address = Undefined
    type BillingAmount = Undefined
    type OrderPlaced = PricedOrder

    type BillableOrderPlaced =
        { OrderId: OrderId
          BillingAddress: Address
          AmountToBill: BillingAmount }

    type PlaceOrderResult =
        { OrderPlaced: OrderPlaced
          BillableOrderPlaced: BillableOrderPlaced
          OrderAcknowledgmentSent: OrderAcknowledgmentSent }

    type PlaceOrderEvent =
        | OrderPlaced of OrderPlaced
        | BillableOrderPlaced of BillableOrderPlaced
        | AcknowledgmentSent of OrderAcknowledgmentSent

    type CreateEvents = PricedOrder -> PlaceOrderEvent list

File: Scala/chap07/src/main/scala/Chap0704.scala
object Chap070401:
  case class ProductCode(value: Any)

  case class ValidatedOrder(value: Any)

  type CheckProductCodeExists = ProductCode => Boolean


  case class UnvalidatedAddress(value: Any)

  case class CheckedAddress(value: UnvalidatedAddress)

  case class AddressValidationError(value: String)

  type CheckAddressExists = UnvalidatedAddress => Either[AddressValidationError, CheckedAddress]

  case class ValidationError(value: Any)

  case class UnvalidatedOrder(value: Any)

  // 書けるけど、こういうやり方はしないような気がする
  // 他にいい方法はあるだろうか...
  // trait 辺りを使うとか
  type ValidateOrder =
    CheckAddressExists // 依存
      => CheckAddressExists // 依存
      => UnvalidatedOrder // 入力
      => Either[ValidationError, ValidatedOrder] // 出力

object Chap070402:

  import Chap070401.*

  case class Price(value: Any)

  case class PricedOrder(value: Any)

  type GetProductPrice = ProductCode => Price

  type PriceOrder = GetProductPrice => ValidateOrder => PricedOrder

object Chap070403:

  import Chap070402.*

  case class EmailAddress(value: Any)

  case class HtmlString(value: String)

  case class OrderAcknowledgment(emailAddress: EmailAddress, letter: HtmlString)

  type CreateOrderAcknowledgmentLetter = PricedOrder => HtmlString

  type SenOrderAcknowledgmentRetVolt = OrderAcknowledgment => Unit

  type SendOrderAcknowledgmentRetBool = OrderAcknowledgment => Boolean

  enum SendResult:
    case Sent
    case NotSent

  type SendOrderAcknowledgmentRetRes = OrderAcknowledgment => SendResult

  case class OrderId(value: Any)

  case class OrderAcknowledgmentSent(orderId: OrderId, emailAddress: EmailAddress)

  type SendOrderAcknowledgementRetEvent = OrderAcknowledgmentSent => Option[OrderAcknowledgmentSent]

  type AcknowledgeOrder =
    CreateOrderAcknowledgmentLetter // 依存
      => SendOrderAcknowledgmentRetRes // 依存
      => PricedOrder // 入力
      => Option[OrderAcknowledgmentSent] // 出力

object Chap070404:

  import Chap0702.PricedOrder
  import Chap070403.{OrderId, OrderAcknowledgmentSent}

  case class Address(value: Any)

  case class BillingAmount(value: Any)

  case class OrderPlaced(value: PricedOrder)

  case class BillableOrderPlaced(orderId: OrderId, billingAddress: Address, amountToBill: BillingAmount)

  case class PlaceOrderResult(
                               orderPlaced: OrderPlaced,
                               billableOrderPlaced: BillableOrderPlaced,
                               orderAcknowledgmentSent: Option[OrderAcknowledgmentSent],
                             )

  enum PlaceOrderEvent:
    case Placed(value: OrderPlaced)
    case BillablePlaced(value: BillableOrderPlaced)
    case AcknowledgmentSent(value: OrderAcknowledgmentSent)

  type CreateEvents = PricedOrder => List[PlaceOrderEvent]